sapirbitan/assignment2/ex1.cpp

```diff
@@ -0,0 +1,126 @@
+#include <iostream>
+using namespace std;
+
+struct node {
+	int key;
+	node* left_child = NULL;
+	node* right_child = NULL;
+	node* ancestor = NULL;
+};//assuming every node contains pointers to it's children and ancestors
+  //(the root's ancestor will point to NULL)
+
+// this function finds the key which we want to print it's ancestors from
+//it goes over all of the tree and returns the node we were looking for
+node* find_key(node& root, int& key) {
+	node* found_key = NULL;
+	if (&root == NULL) 
+		return NULL	;
+	if (root.key == key)
+		return &root;
+	//since it's a binary tree - we have to go in two paths - left and right
+	found_key = find_key(*(root.left_child), key);//left
+	if (found_key != NULL)
+		return found_key;
+	found_key = find_key(*(root.right_child), key);//right
+	if (found_key != NULL)
+		return found_key;
+	return NULL;// if we haven't found anything - return NULL
+}
+
+//this function prints the key's ancestors
+void print_anc(node& key){
+	if (key.ancestor == NULL) {//since the only node that has no ancestor is the root
+		cout << "root does not have ancestors" << endl;
+	}
+	node current_node = key;
+	while (current_node.ancestor != NULL) {
+		cout << current_node.ancestor->key<< "\n";
+		current_node = *current_node.ancestor;
+	}
+
+}
+void testing(node& root) {
+	//test1 - should print - 2 3 7
+	int num = 6;
+	node* key = find_key(root, num);
+	cout << "test 1: key = " <<num<< endl;
+	if (key != NULL)
+		print_anc(*key);
+	else
+		cout << "the key does not exist in this tree";
+	//test2 - should print - 3 7
+	num = 5;
+	cout << "test 2: key = " << num << endl;
+	key = find_key(root, num);
+	if (key != NULL)
+		print_anc(*key);
+	else
+		cout << "the key does not exist in this tree";
+	//test3 - should print - 4 7
+	num = 8;
+	cout << "test 3: key = " << num << endl;
+	key = find_key(root, num);
+	if (key != NULL)
+		print_anc(*key);
+	else
+		cout << "the key does not exist in this tree";
+	//test4 - should print "root does not have ancestors"
+	num = 7;
+	cout << "test 4: key = " << num << endl;
+	key = find_key(root, num);
+	if (key != NULL)
+		print_anc(*key);
+	else
+		cout << "the key does not exist in this tree";
+	//test5 - should print - 2 3 7
+	num = 1;
+	cout << "test 5: key = " << num << endl;
+	key = find_key(root, num);
+	if (key != NULL)
+		print_anc(*key);
+	else
+		cout << "the key does not exist in this tree";
+	//test6 - should print - "the key does not exist in this tree"
+	num = 13;
+	cout << "test 6: key = " << num << endl;
+	key = find_key(root, num);
+	if (key != NULL)
+		print_anc(*key);
+	else
+		cout << "the key does not exist in this tree";
+}
+
+int main() {
+	node root;
```

Comment:
When dealing with data structures that might be difficult to understand or visualise, it's helpful to 'draw' out the structure. Consider adding a comment with ASCII representation of your tree:

```
          5
       /    \
      3      6
     /  \      \
   11   1       2
```